materialize-webhook/driver.go

```diff
@@ -161,87 +197,107 @@ func (d *transactor) Load(it *m.LoadIterator, _ func(int, json.RawMessage) error
 	return nil
 }
 
-// Store invokes the Webhook URL, with a body containing StoreIterator documents.
+// Store streams StoreIterator documents directly into the webhook body.
 func (d *transactor) Store(it *m.StoreIterator) (m.StartCommitFunc, error) {
-	var bodies = make([]bytes.Buffer, len(d.addresses))
-	var ctx = it.Context()
+	var pipeWriter *io.PipeWriter
+	group := errgroup.Group{}
 
-	// TODO(johnny): perform incremental POSTs rather than queuing a single call.
-	for it.Next() {
-		var b = &bodies[it.Binding]
+	startWebhook := func(address string) {
+		pipeReader, w := io.Pipe()
 
-		if b.Len() != 0 {
-			b.WriteString(",\n")
-		} else {
-			b.WriteString("[\n")
-		}
-		if _, err := b.Write(it.RawJSON); err != nil {
-			return nil, err
-		}
-	}
-
-	for i := range bodies {
-		bodies[i].WriteString("\n]")
-	}
-
-	for i, address := range d.addresses {
-		var address = address.String()
-		var body = &bodies[i]
-
-		for attempt := 0; true; attempt++ {
-			select {
-			case <-ctx.Done():
-				return nil, ctx.Err()
-			case <-time.After(backoff(attempt)):
-				// Fallthrough.
-			}
-
-			request, err := http.NewRequest("POST", address, bytes.NewReader(body.Bytes()))
+		group.Go(func() error {
+			request, err := http.NewRequest("POST", address, pipeReader)
 			if err != nil {
-				return nil, fmt.Errorf("http.NewRequest(%s): %w", address, err)
+				return fmt.Errorf("http.NewRequest(%s): %w", address, err)
 			}
+
 			request.Header.Add("Content-Type", "application/json")
 
+			for i := range d.customHeaders {
+				header := d.customHeaders[i]
+				request.Header.Add(header.Name, header.Value)
+			}
+
 			response, err := http.DefaultClient.Do(request)
+
 			if err == nil {
 				err = response.Body.Close()
 			}
+
 			if err == nil && (response.StatusCode < 200 || response.StatusCode >= 300) {
```

Comment:
I think there needs to be a check for `if err != nil` somewhere in here. It's also important to consider what happens if this goroutine errors before we call `wait` on the errgroup - writes to the writer end of the pipe will block forever since there is nothing left reading from the reader end.

You should add some handling to close the read end if the reading process errors, so that the error will be propogated to `write` calls. See https://github.com/estuary/connectors/blob/a492d4573331be75a0ad55e7ea3b8ac1fbec4611/materialize-redshift/staged_file.go#L111-L124 for an example of this. You'll probably ened to do it if `http.NewRequest` fails, or if `http.DefaultClient.Do` fails.

Reply:
Got it. I used `CloseWithError` in both of those cases plus if `response.Body.Close` fails or if we receive a non-2xx response, so those errors can also be propagated to the `write` calls.